File: sdk-go/signals.go
package primitive

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
	"unicode/utf16"
	"unicode/utf8"
)

// SignalParent contains explicit caller-authenticated addressing and account context.
type SignalParent struct {
	AccountScope string   `json:"accountScope"`
	From         string   `json:"from"`
	To           string   `json:"to"`
	MessageID    *string  `json:"messageId"`
	Subject      string   `json:"subject"`
	References   []string `json:"references"`
}
type SignalInput struct {
	Parent      SignalParent `json:"parent"`
	Kind        string       `json:"kind"`
	Status      string       `json:"status,omitempty"`
	Note        *string      `json:"note,omitempty"`
	ExpiresAtMs *int64       `json:"expiresAtMs,omitempty"`
}
type SignalDependencies struct {
	UUID func() string
	// Now returns Unix time in integer milliseconds.
	Now func() int64
}

// PreparedSignal is a value snapshot. Persist all fields before dispatch; do not
// change fields between retries. RequestJSON fixes the ordinary send body.
type PreparedSignal struct {
	AccountScope   string `json:"accountScope"`
	PreparedAtMs   int64  `json:"preparedAtMs"`
	ExpiresAtMs    *int64 `json:"expiresAtMs"`
	IdempotencyKey string `json:"idempotencyKey"`
	RequestJSON    string `json:"requestJson"`
}
type SignalPreparation struct {
	Status   string          `json:"status"`
	Prepared *PreparedSignal `json:"prepared,omitempty"`
}

// SignalSendResult preserves the ordinary operation's response. An expired
// result is a local refusal, never evidence that an earlier attempt failed.
type SignalSendResult[T any] struct {
	Status         string
	Result         T
	IdempotencyKey string
}

var signalUUID = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$`)
var signalMailbox = regexp.MustCompile("^[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+)*@[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?(?:\\.[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?)*$")
var signalWireID = regexp.MustCompile(`^[!-~]+@[!-~]+$`)

func signalHeader(value string, limit int) bool {
	if !utf8.ValidString(value) || len(value) > limit {
		return false
	}
	for _, c := range value {
		if c < 32 || c == 127 {
			return false
		}
	}
	return true
}
func signalMessageID(value string) (string, error) {
	value = strings.Trim(value, " ")
	if strings.HasPrefix(value, "<") && strings.HasSuffix(value, ">") {
		value = value[1 : len(value)-1]
	}
	if !signalWireID.MatchString(value) || strings.ContainsAny(value, "<>") || strings.Count(value, "@") != 1 || len(value) > 996 {
		return "", fmt.Errorf("invalid Message-ID")
	}
	return "<" + value + ">", nil
}
func signalMilliseconds(value int64) bool { return value >= 0 && value <= 253402300799999 }

// signalJSON disables HTML escaping so the UTF-8 attachment is identical across SDKs.
func signalJSON(value any) (string, error) {
	var out strings.Builder
	encoder := json.NewEncoder(&out)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(value); err != nil {
		return "", err
	}
	return strings.TrimSuffix(out.String(), "\n"), nil
}

// PrepareSignalEmail builds one explicit ordinary email without performing IO.
func PrepareSignalEmail(input SignalInput, dependencies SignalDependencies) (SignalPreparation, error) {
	var zero SignalPreparation
	p := input.Parent
	if p.MessageID == nil || *p.MessageID == "" {
		return SignalPreparation{Status: "waiting_on_parent"}, nil
	}
	target, err := signalMessageID(*p.MessageID)
	if err != nil {
		return zero, err
	}
	for _, address := range []string{p.From, p.To} {
		if !signalHeader(address, 320) || !signalMailbox.MatchString(address) {
			return zero, fmt.Errorf("use one bare mailbox per address")
		}
	}
	if !signalHeader(p.Subject, 998) || !signalHeader(p.AccountScope, 256) || p.AccountScope == "" {
		return zero, fmt.Errorf("invalid subject or account scope")
	}
	if len(p.References) > 1000 {
		return zero, fmt.Errorf("too many references")
	}
	references := make([]string, 0, len(p.References)+1)
	for _, value := range p.References {
		ref, err := signalMessageID(value)
		if err != nil {
			return zero, err
		}
		if ref != target {
			references = append(references, ref)
		}
	}
	references = append(references, target)
	length := -1
	for _, ref := range references {
		length += len(ref) + 1
	}
	start := 0
	for len(references)-start > 100 || length > 8192 {
		length -= len(references[start]) + 1
		start++
	}
	references = references[start:]
	if p.Subject == "" {
		p.Subject = "Re: Your message"
	}
	if dependencies.Now == nil || dependencies.UUID == nil {
		return zero, fmt.Errorf("clock and UUID callbacks are required")
	}
	now := dependencies.Now()
	if !signalMilliseconds(now) {
		return zero, fmt.Errorf("invalid clock")
	}
	var expires *int64
	var expiry *string
	payload := struct {
		SubjectMessageID string  `json:"subject_message_id"`
		Status           string  `json:"status,omitempty"`
		Note             *string `json:"note,omitempty"`
	}{SubjectMessageID: target}
	switch input.Kind {
	case "ack":
		if signalText("ack", input.Status, nil) == "" {
			return zero, fmt.Errorf("invalid ACK status")
		}
		payload.Status = input.Status
		if input.Note != nil {
			if !utf8.ValidString(*input.Note) || len(utf16.Encode([]rune(*input.Note))) > 2000 || strings.ContainsRune(*input.Note, 0) {
				return zero, fmt.Errorf("invalid note")
			}
			note := *input.Note
			payload.Note = &note
		}
	case "read":
	case "working":
		if input.ExpiresAtMs == nil || !signalMilliseconds(*input.ExpiresAtMs) || *input.ExpiresAtMs <= now || *input.ExpiresAtMs-now > 60000 {
			return zero, fmt.Errorf("working expiry must be within 60 seconds")
		}
		value := *input.ExpiresAtMs
		expires = &value
		formatted := time.UnixMilli(value).UTC().Format("2006-01-02T15:04:05.000Z")
		expiry = &formatted
	default:
		return zero, fmt.Errorf("invalid signal kind")
	}
	interaction, step := strings.ToLower(dependencies.UUID()), strings.ToLower(dependencies.UUID())
	if !signalUUID.MatchString(interaction) || !signalUUID.MatchString(step) || interaction == step {
		return zero, fmt.Errorf("two distinct UUIDs are required")
	}
	envelope := struct {
		Version         int     `json:"interaction_version"`
		ID              string  `json:"interaction_id"`
		Protocol        string  `json:"protocol"`
		ProtocolVersion int     `json:"protocol_version"`
		Step            string  `json:"step"`
		StepID          string  `json:"step_id"`
		Prev            *string `json:"prev_step_id"`
		Expires         *string `json:"expires_at"`
		Payload         any     `json:"payload"`
	}{1, interaction + "@" + strings.Split(p.To, "@")[1], input.Kind, 1, input.Kind, step, nil, expiry, payload}
	encoded, err := signalJSON(envelope)
	if err != nil {
		return zero, err
	}
	if len(encoded) > 65536 {
		return zero, fmt.Errorf("signal exceeds 64 KiB")
	}
	attachment := struct {
		Filename      string `json:"filename"`
		ContentType   string `json:"content_type"`
		ContentBase64 string `json:"content_base64"`
	}{"interaction.json", "application/json", base64.StdEncoding.EncodeToString([]byte(encoded))}
	body := struct {
		From        string   `json:"from"`
		To          string   `json:"to"`
		Subject     string   `json:"subject"`
		BodyText    string   `json:"body_text"`
		InReplyTo   string   `json:"in_reply_to"`
		References  []string `json:"references"`
		Attachments any      `json:"attachments"`
	}{p.To, p.From, p.Subject, signalText(input.Kind, input.Status, input.Note), target, references, []any{attachment}}
	request, err := signalJSON(body)
	if err != nil {
		return zero, err
	}
	return SignalPreparation{Status: "prepared", Prepared: &PreparedSignal{p.AccountScope, now, expires, "signal-" + step, request}}, nil
}

// SendPreparedSignal makes one ordinary attempt using the supplied operation.
// The fresh RawMessage copy cannot alter the saved request for subsequent retries.
func SendPreparedSignal[T any](ctx context.Context, sendMail func(context.Context, json.RawMessage, string) (T, error), prepared PreparedSignal, accountScope string, now func() int64) (SignalSendResult[T], error) {
	var zero SignalSendResult[T]
	if accountScope == "" || accountScope != prepared.AccountScope {
		return zero, fmt.Errorf("account scope mismatch")
	}
	if now == nil {
		return zero, fmt.Errorf("clock is required")
	}
	observed := now()
	if !signalMilliseconds(observed) {
		return zero, fmt.Errorf("invalid clock")
	}
	if prepared.ExpiresAtMs != nil && observed >= *prepared.ExpiresAtMs {
		return SignalSendResult[T]{Status: "expired", IdempotencyKey: prepared.IdempotencyKey}, nil
	}
	result, err := sendMail(ctx, json.RawMessage(prepared.RequestJSON), prepared.IdempotencyKey)
	if err != nil {
		return zero, err
	}
	return SignalSendResult[T]{Status: "response", Result: result}, nil
}

func signalText(kind, status string, note *string) string {
	if kind == "read" {
		return "I read your message."
	}
	if kind == "working" {
		return "I am working on your message."
	}
	text := map[string]string{"received": "Received your message.", "will_process": "I intend to process your message.", "will_not_process": "I will not process your message."}[status]
	if note != nil {
		text += "\n\n" + *note
	}
	return text
}

File: sdk-python/src/primitive/signals.py
"""Explicit optional email signals; no automatic emission or persistence."""

import base64
import json
import re
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Generic, Literal, TypeVar


@dataclass(frozen=True)
class SignalParent:
    account_scope: str
    from_email: str
    to: str
    message_id: str | None
    subject: str | None
    references: tuple[str, ...]


@dataclass(frozen=True)
class SignalInput:
    parent: SignalParent
    kind: Literal["ack", "read", "working", "typing"]
    status: Literal["received", "will_process", "will_not_process"] | None = None
    note: str | None = None
    expires_at_ms: int | None = None


@dataclass(frozen=True)
class PreparedSignal:
    """Persist all fields before dispatch. request_json fixes the send body."""

    account_scope: str
    prepared_at_ms: int
    expires_at_ms: int | None
    idempotency_key: str
    request_json: str


@dataclass(frozen=True)
class SignalPreparation:
    status: Literal["waiting_on_parent", "prepared"]
    prepared: PreparedSignal | None = None


@dataclass(frozen=True)
class ExpiredSignal:
    """Local refusal only. Reconcile the saved key if any attempt may have run."""

    idempotency_key: str
    status: Literal["expired"] = "expired"


_UUID = re.compile(
    r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
)
_ATOM = r"[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+"
_LABEL = r"[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?"
_MAILBOX = re.compile(rf"{_ATOM}(?:\.{_ATOM})*@{_LABEL}(?:\.{_LABEL})*")
_BODIES = {
    "received": "Received your message.",
    "will_process": "I intend to process your message.",
    "will_not_process": "I will not process your message.",
}


def _check(condition: bool, message: str) -> None:
    if not condition:
        raise ValueError(message)


def _header(value: str, limit: int) -> None:
    _check(
        not re.search(r"[\x00-\x1f\x7f]", value)
        and len(value.encode("utf-8")) <= limit,
        "invalid header",
    )


def _wire_id(value: str) -> str:
    value = value.strip(" ")
    if value.startswith("<") and value.endswith(">"):
        value = value[1:-1]
    _check(
        bool(re.fullmatch(r"[!-~]+@[!-~]+", value))
        and not re.search(r"[<>]", value)
        and value.count("@") == 1
        and len(value) <= 996,
        "invalid Message-ID",
    )
    return f"<{value}>"


def _json(value: object) -> str:
    return (
        json.dumps(value, ensure_ascii=False, separators=(",", ":"))
        .replace("\u2028", "\\u2028")
        .replace("\u2029", "\\u2029")
    )


def _milliseconds(value: int) -> None:
    _check(
        type(value) is int and 0 <= value <= 253402300799999, "invalid clock or expiry"
    )


def prepare_signal_email(
    signal: SignalInput, *, uuid: Callable[[], str], now: Callable[[], int]
) -> SignalPreparation:
    parent = signal.parent
    if parent.message_id is None or parent.message_id == "":
        return SignalPreparation("waiting_on_parent")
    target = _wire_id(parent.message_id)
    for address in (parent.from_email, parent.to):
        _header(address, 320)
        _check(bool(_MAILBOX.fullmatch(address)), "use one bare mailbox per address")
    subject = parent.subject or ""
    _header(subject, 998)
    _header(parent.account_scope, 256)
    _check(bool(parent.account_scope), "account_scope is required")
    _check(len(parent.references) <= 1000, "too many references")
    references = [
        ref for value in parent.references if (ref := _wire_id(value)) != target
    ]
    references.append(target)
    length = sum(len(ref) + 1 for ref in references) - 1
    start = 0
    while len(references) - start > 100 or length > 8192:
        length -= len(references[start]) + 1
        start += 1
    references = references[start:]
    observed = now()
    _milliseconds(observed)
    expires = None
    payload = {"subject_message_id": target}
    if signal.kind == "ack":
        _check(signal.status in _BODIES, "invalid ACK status")
        assert signal.status is not None
        payload["status"] = signal.status
        if signal.note is not None:
            _check(
                len(signal.note.encode("utf-16-le")) <= 4000
                and "\0" not in signal.note,
                "note exceeds 2000 UTF-16 units",
            )
            payload["note"] = signal.note
    elif signal.kind != "read":
        _check(signal.kind in ("working", "typing"), "invalid signal kind")
        _check(signal.expires_at_ms is not None, f"{signal.kind} expiry is required")
        assert signal.expires_at_ms is not None
        _milliseconds(signal.expires_at_ms)
        _check(
            0 < signal.expires_at_ms - observed <= 60000,
            f"{signal.kind} expiry must be within 60 seconds",
        )
        expires = signal.expires_at_ms
    interaction, step = uuid().lower(), uuid().lower()
    _check(
        bool(_UUID.fullmatch(interaction))
        and bool(_UUID.fullmatch(step))
        and interaction != step,
        "two distinct UUIDs are required",
    )
    envelope = {
        "interaction_version": 1,
        "interaction_id": f"{interaction}@{parent.to.split('@')[1]}",
        "protocol": signal.kind,
        "protocol_version": 1,
        "step": signal.kind,
        "step_id": step,
        "prev_step_id": None,
        "expires_at": None
        if expires is None
        else datetime.fromtimestamp(expires // 1000, timezone.utc)
        .replace(microsecond=expires % 1000 * 1000)
        .isoformat(timespec="milliseconds")
        .replace("+00:00", "Z"),
        "payload": payload,
    }
    encoded = _json(envelope).encode("utf-8")
    _check(len(encoded) <= 65536, "signal exceeds 64 KiB")
    body = {
        "from": parent.to,
        "to": parent.from_email,
        "subject": subject or "Re: Your message",
        "body_text": _signal_text(signal.kind, signal.status, signal.note),
        "in_reply_to": target,
        "references": references,
        "attachments": [
            {
                "filename": "interaction.json",
                "content_type": "application/json",
                "content_base64": base64.b64encode(encoded).decode("ascii"),
            }
        ],
    }
    return SignalPreparation(
        "prepared",
        PreparedSignal(
            parent.account_scope, observed, expires, f"signal-{step}", _json(body)
        ),
    )


T = TypeVar("T")


@dataclass(frozen=True)
class SignalResponse(Generic[T]):
    result: T
    status: Literal["response"] = "response"


async def send_prepared_signal(
    send_mail: Callable[[dict[str, object], str], Awaitable[T]],
    prepared: PreparedSignal,
    *,
    account_scope: str,
    now: Callable[[], int],
) -> SignalResponse[T] | ExpiredSignal:
    """Check scope and expiry when awaited, then make one ordinary send attempt.

    A fresh request copy protects subsequent retries. Exceptions propagate.
    """
    _check(
        bool(account_scope) and account_scope == prepared.account_scope,
        "account scope mismatch",
    )
    observed = now()
    _milliseconds(observed)
    if prepared.expires_at_ms is not None and observed >= prepared.expires_at_ms:
        return ExpiredSignal(prepared.idempotency_key)
    return SignalResponse(
        await send_mail(json.loads(prepared.request_json), prepared.idempotency_key)
    )


def _signal_text(kind: str, status: str | None = None, note: str | None = None) -> str:
    if kind == "read":
        return "I read your message."
    if kind == "working":
        return "I am working on your message."
    if kind == "typing":
        return "I am composing a reply to your message."
    text = _BODIES.get(status or "", "")
    return text if note is None else text + "\n\n" + note
